src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,156 @@
+| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |
+
+3. Commit all secondaries. Finally, we get this:
+
+| key | bal:data | bal:lock | bal:write |
+|-----|----------|----------|-----------|
+| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+| Joe | 8:<br/>7:$9<br/>6:<br/>5:$2 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+
+### Reading
+
+To read from Percolator, a timestamp is required too. The procedure to perform read operation is as following:
+
+1. Get a timestamp, `ts`.
+2. Check if the row we are going to read is locked with a timestamp in the range `[0, ts]`;
+    * If so, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before `ts` or after `ts`. Other wise, we can continue reading.
+3. Get the latest record in the row's `write` column whose `commit_ts` is in range `[0, ts]`. The record contains the `start_ts` of the transaction where it was committed.
+4. Get the row's value in `data` column, whose timestamp is exactly `start_ts`. Then the value is what we want.
+
+For example, consider this table again:
+
+| key | bal:data | bal:lock | bal:write |
+|-----|----------|----------|-----------|
+| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: |8:data@7<br/>7:<br/>6:data@5<br/>5: |
+| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |
+
+Let's read Bob's balance.
+1. Get a timestamp. Assume we got `9`.
+2. Check lock. The row of Bob is not locked, so we continue.
+3. Get the latest record in `write` column that committed before `9`. We got the record with `commit_ts` equals to `8`, and it says the `start_ts` is `7`, which means, it's corresponding data is at timestamp `7` in `data` column.
+4. Get the value in `data` column whose timestamp is exactly `7`. We got `$3`.
+
+This algorithm provides us both lock-free reading and the ability to perform historical read. In the above example, if we specify that we want to read at time point `7`, then we will see the write record at timestamp `6`, giving us the result `$10` at timestamp `5`.
+
+### Handling Conflicts
+
+Conflicts are identified by checking the `lock` column. A row can have many versions of data, but it can have at most one lock at any time.
+
+When we are performing a write operation, we try to lock every affected row in the `Prewrite` phase. If we failed to lock some of these rows, the whole transaction will be rolled back. As an optimistic lock algorithm, sometimes it causes a shortage of Percolator: there may be performance regression in a scenario that conflicts occur frequently.
+
+To roll back a row, just simply remove its lock and its corresponding value in `data` column.
+
+### Tolerating crashes
+
+Percolator has the ability to survive crashes without breaking the correctness of data.
```

Comment:
```suggestion
Percolator has the ability to survive crashes without compromising data integrity.
```

Reply:
Can I use "breaking" instead of "compromising"?

Reply:
yeah it is fine

Reply:
Sure. It's just a personal preference. Both are correct.